### DISCUSSION TITLE
Target-typed collection literals

### DISCUSSION THREAD

**Discussion by Richiban** [dotnet/csharplang#1399] [Language Ideas] [answered]
C# is in dire need of some collection literals, since even the shortest version available is still quite painful:

```csharp
var data = new [] {1, 2, 3, 4, 5};
```

(A shorter syntax does exist, but only in certain circumstances):

```csharp
int[] data = {1, 2, 3, 4, 5};
```

I propose using the very widely used `[x, y, z]` syntax for collections, but instead of being useful only for arrays / lists / whatever the chosen master collection of the language is, I propose that they be target-typed.

If no target-type is available, we fall back to array. Might Span<T> be a better choice? Feels like immutability would be appropriate here.

E.g.

```csharp
var nums = [1, 2, 3]; // No target type, so int[] is chosen
```

```csharp
List<string> numbers = ["one", "two", "three"]; // numbers is obviously of type List<string>
```

How is the collection constructed (for unknown types)?

It should be duck-typed (or the static equivalent) based on the following possible rules:
    1. a constructor taking an `IEnumerable<T>`
    2. a constructor taking `T[]`
    3. a parameterless constructor and a `void Add(T)` method
    4. an explicit or implicit cast from `T[]`
    5. a specially-named new method (perhaps `Create(T[])`?) in the vein of the new `Deconstruct` feature

From the list above it seems that (3) is the most C#-ish, since that would marry with how collection initialisers currently work. However, it's likely to be less useful than the others since it's inherently incompatible with immutable types.

I think that for any constructors or methods `T[]` should be ruled out in favour of `IEnumerable<T>` to avoid the user from having to potentially write the method or constructor twice (motivation for this should be the same as in https://github.com/dotnet/csharplang/issues/179#issuecomment-288488262).

The main advantage of (1) is that there an awful lot of types--both in the BCL and in the community--that already have a constructor with this signature and are therefore coincidentally compatible with this new feature.

The main advantage of (5) in the case of `Deconstruct` is that extension methods can be used to add the funcionality to types that the author does not own. Since the `Create(T[])` method in this feature would be static unfortunately extension methods are not going to help us.

Is it possible, or even desireable, to have some duck-typing that would look for a constructor and, if one is not found, move on to looking for a `Create` method?

This feature has links to "Target-typed `new` expression" https://github.com/dotnet/csharplang/issues/100 and "Dictionary Literals" https://github.com/dotnet/csharplang/issues/414.

Next up for discussion: collection patterns. They've been largely undiscussed until now because it seems crazy to support quick *deconstruction* of collections but not *construction*.

**Comment by svick** [comment]
What if the method was actually `Create(ReadOnlySpan<T>)`? That way, the collection could be created without additional allocations.

**Comment by Richiban** [comment]
@svick I did consider `Span<T>` and `ReadOnlySpan<T>`, but liked the `IEnumerable<T>` approach simply because there are all these types out there that will already be compatible with this feature.

`ReadOnlySpan<T>` does seem the most *correct* type to use though.

**Comment by jveselka** [comment]
Not saying that this is bad proposal, but do we really need yet another collection initialization syntax? I think they have already beaten delegates in how many ways you can write the same thing. :-)

**Comment by Richiban** [comment]
@zippec 

> do we really need yet another collection initialization syntax?

Well, you _are_ right. Unfortunately though all the existing methods are **bad**. There's little consistency between them and they differ in when they can be used which just creates a porridge of confusion for new and even intermediate users.

E.g.

```csharp
int[] data = {1, 2, 3}; // Only works when in a variable initialisation statement
                        // Cannot use type inference
                        // Only works on arrays.
```

```csharp
var data = new [] {1, 2, 3}; // Works in any expression context, but only for arrays
                             // Slightly strange syntax
                             // Works with type inference.
```

```csharp
var data = new List<int> {1, 2, 3}; // Works in any expression context
                                    // Only works on *mutable* collection types
                                    // Does not work with type inference.
```

And now, my proposal:


```csharp
var data = [1, 2, 3]; // Works in any expression context
                      // for any collection type
                      // works with type inference.
```

---
Note: 
When I say "does not work with type inference", I mean that initialising the collection requires explicitly stating the generic type parameter. For example, in C# it's not possible to define a `List` of anonymous objects:

```csharp
//  There's no type I can write here 
//                  ↓
var data = new List< > { new { x = 1 } };

// My only choice is this, which involves an array copy:
var data = new [] { new { x = 1 } }.ToList();
```

As an aside, when target-typed `new` arrives, at least we'll be able to write this:
```csharp
List<object> items = new(new [] { new { x = 1 } });
```

:grin:

**Comment by quinmars** [comment]
I think with target-typed `new`, it'd be possible to write:

```csharp
List<int> items = new () {1, 2, 3};
```

**Comment by MgSam** [comment]
>C# is in dire need of some collection literals...

I guess that depends on your definition of "dire" :)

Why is `List<string> numbers = ["one", "two", "three"];` better than `var numbers = new List<string>{"one", "two", "three"};`?

**Comment by Richiban** [comment]
@MgSam 

> Why is List<string> numbers = ["one", "two", "three"]; better than var numbers = new List<string>{"one", "two", "three"};

In that particular instance it's six of one and half a dozen of the other, but in other situations it leads to a noticeable improvement:

```csharp
var xs = [1, 2, 3];
```

```csharp
public void M(List<int> xs)
{
    ...
}

M([1, 2, 3]);
```

```csharp
class O 
{
    public ImmutableList<int> Prop { get; set; }
}

o.Prop = [1, 2, 3];
```

**Comment by bjorn-ali-goransson** [comment]
Please, immutables as fallback... 🤲

Lately, I've been contemplating the potential security risks of passing and most prominently returning IEnumerable<...>, as if being immutable when it is actually mutable and possibly (but not probably) a shared instance, at least in scope.

For example:

```
List<int> GetNumbersFromXmlFile()
{
    return new List<int> { 1, 2, 3 };
}

public IEnumerable<int> GetNumbers()
{
    var n = GetNumbersFromXmlFile();
    LogStuff(n);
    return n;
}

void LogStuff(IEnumerable<int> numbers){
    ((List<int>)numbers)[0] = -1; 👌
}
```

An admittedly improbable but still not impossible, and if used cleverly, very powerful attack vector that is overlooked by virtually everyone and anyone.

Also, I've thought about the stylish Enumerable.Empty<...>() which is really nice and a whole lot better (and more secure) than returning an empty List posing as an IEnumerable. One should really be able to create an anonymous implementation of the IEnumerable more easily. This proposition would fill that need!

As it is now, all my methods returning IEnumerable<...> now return `result.ToList().AsReadOnly()` (which is repetitive and error-prone) even though result is itself already a List, lest someone uses that object unresponsibly.

Thumbs up to this.

**Comment by surfsky** [comment]
Any progress with this issue?
It' s only a simply, readable syntax, which is used by so many code langs.

``` csharp
var items = [1, 2, 3];           // create List<int>{.....}
var items = ["a", "b", "c"];     // create List<string>{....}
var items = {                    // create dictionary<string, int>
    "key1":  1,
    "key2": 2,
    "key3": 3
};
```

**Comment by theunrepentantgeek** [comment]
```
var items = [1, 2, 3]; 
```
Why should this syntax return a `List<int>` and not an `int[]`, `Queue<int>`, `Stack<int>` or an `ImmutableArray<int>` ?

This is the biggest problem with introducing syntax like this - elevating one particular type as better than all the others, with the inevitable consequence of promoting that type for use in situations where it's not the best choice.

Other languages that have syntax like this suffer from this problem. Go, for example, pushes everything into being a slice - so there are slices used as collections, slices used as stacks, slices used as queues, slices used as heaps, and so on. Everything is forced into the same type by the shape of the language.

**Comment by bjorn-ali-goransson** [comment]
@theunrepentantgeek Lists are the more versatile and widely used of all those candidates. That is why many propose Lists.

For me, I would be fine with an array as well, if it's totally mutable. We currently don't have any good way of instantiating immutable enumerables. new List<int>{...} isn't very safe, and .AsReadOnly() seems overly wasteful.

Are perhaps any of the new data types usable in this context? (dont know)

**Comment by jnm2** [comment, marked as answer]
Good news! This is championed at https://github.com/dotnet/csharplang/issues/5354 and has a working group.

> Next up for discussion: collection patterns. They've been largely undiscussed until now because it seems crazy to support quick deconstruction of collections but not construction.

List patterns shipped in C# 11! https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#list-patterns
